### PR TITLE
[SYCL][CUDA] Ignore cuda prefetch hint if not supported

### PR DESCRIPTION
**Pull request by AidanBeltonS**
Specific devices and OS's, like Windows, do not support concurrent managed memory. `cudaPrefetchAsync` requires concurrent managed access for unified memory. This PR removes the windows error message and replaces it with a check for concurrent managed access. As the SYCL prefetch operation is a hint, this can return a success.

Let me know if there is a preferred error code to throw. Also, if it is best that a user warning is printed to indicate that the hint is being ignored as the device does not support the operation.

### CODE REVIEW

File: sycl/plugins/cuda/pi_cuda.cpp
Comment (steffenlarsen) [inline sycl/plugins/cuda/pi_cuda.cpp:4663]: I think a warning would make sense, but I wonder what the best way to do it would be. I am not sure printing is the best option - at least not from the PI implementation. Maybe a better solution is to add a new "error" code, such as `PI_UNSUPPORTED_OPERATION`, then the runtime can handle it accordingly.

On the other hand, prefetch is a "hint" so ignoring the operation entirely when we don't support it is arguably still valid.

Comment (romanovvlad) [reply sycl/plugins/cuda/pi_cuda.cpp:4663]: +1 for ignoring a hint if it cannot be used/not yet implemented.

For letting user know what happens we could use SYCL_RT_WARNING_LEVEL which is being introduced here https://github.com/intel/llvm/pull/4918

Comment (steffenlarsen) [reply sycl/plugins/cuda/pi_cuda.cpp:4663]: Would `SYCL_RT_WARNING_LEVEL` be available in PI plugins? Otherwise it would still have to inform the runtime somehow.

Comment (romanovvlad) [reply sycl/plugins/cuda/pi_cuda.cpp:4663]: Interesting question. I tend to say that `SYCL_RT_WARNING_LEVEL` should be available for the whole SYCL stack. Propagating detailed warning(which not only says something is not done, but also says why) to the core SYCL RT could be problematic.
Tagging @v-klochkov

**Comment by AerialMantis** [comment]
@steffenlarsen @romanovvlad @v-klochkov just pinging this one as there's not been much activity for a while, I think we just need to decide how to handle the warning, it was suggested that we use `SYCL_RT_WARNING_LEVEL`, should we wait for #4918 to merge?

**Comment by romanovvlad** [comment]
> 

I would suggest waiting unless there is urgency.
 @v-klochkov Could you please comment on when the patch with SYCL_WARNING_LEVEL is planned to be merged?

**Comment by v-klochkov** [comment]
> > 
> 
> I would suggest waiting unless there is urgency. @v-klochkov Could you please comment on when the patch with SYCL_WARNING_LEVEL is planned to be merged?

GPU RT remains the same after beginning of November'2021. It is still: 21.46.21636. That RT is too old for #4918 . It is good idea to split that patch into 2 parts: a) adding the new env variable and b) adding the switch to VC/GPU backend. Adding (a) would unblock this patch here.

I am on vacation/sabbatical & traveling right now and temporarily cannot split the patch now even though  that is very simple work .

**Comment by AerialMantis** [comment]
Thanks for the update, this isn't urgent so I think it can wait, though I did see that https://github.com/intel/llvm/pull/4918 has been merged now, would this be sufficient to use `SYCL_WARNING_LEVEL` in this patch?

**Comment by bader** [comment]
> #4918 has been merged now

https://github.com/intel/llvm/pull/4918 is still open.

**Comment by AerialMantis** [comment]
Sorry pasted the wrong PR, I meant this one - https://github.com/intel/llvm/pull/5319

File: sycl/plugins/cuda/pi_cuda.cpp
Comment (pvchupin) [reply sycl/plugins/cuda/pi_cuda.cpp:4663]: @smaslov-intel, @v-klochkov, can you comment please?

Comment (smaslov-intel) [reply sycl/plugins/cuda/pi_cuda.cpp:4663]: I think it is OK to re-read the `SYCL_RT_WARNING_LEVEL` from plugins. We already do similar for `SYCL_PI_TRACE` inside Level-Zero plugin: https://github.com/intel/llvm/blob/8213321ebb90110bf4f3d04fa0dc8e131a464a19/sycl/plugins/level_zero/pi_level_zero.cpp#L1822

On the other hand maybe it is a good time to add formal support for backend-specific warnings/errors. Something like returning fixed PI_PLUGIN_SPECIFIC_ERROR, which would indicate that some warning/error is generated and can be obtained by a subsequent call to [new] `piPluginGetLastError`.

**Comment by v-klochkov** [comment]
> Sorry pasted the wrong PR, I meant this one - #5319

Confirmed. #5319 added/merged the new variable SYCL_RT_WARNING_LEVEL. It can be used in this PR.

File: sycl/plugins/cuda/pi_cuda.cpp
Comment (AidanBeltonS) [reply sycl/plugins/cuda/pi_cuda.cpp]: @smaslov-intel I think the `PI_PLUGIN_SPECIFIC_ERROR` approach would seem the most reasonable way to emit a warning/error message. I have implemented it in https://github.com/intel/llvm/pull/5043/commits/478a546a990955e3814a1dfe6522e2a84fced709 for the CUDA backend. Let me know if this approach seems okay and I will extend it to the other backends.

File: sycl/plugins/cuda/pi_cuda.cpp
Comment (romanovvlad) [inline sycl/plugins/cuda/pi_cuda.cpp:4760]: Could you please clarify why can't we return a PI error code which would be interpreted on the SYCL RT side?
How can we guarantee that getLastError approach will correctly work in multithreaded environment?

**Review by romanovvlad — CHANGES_REQUESTED** [review]

File: sycl/plugins/cuda/pi_cuda.cpp
Comment (AidanBeltonS) [reply sycl/plugins/cuda/pi_cuda.cpp:4760]: Sure, I think the main reason it does not make sense to return an error code is that it would be very specific to this backend. This error code would have to express that the operation failed due to the os/device not supporting concurrent memory and that this error is not a fail case but more of a performance warning as there is a redundant prefetch call. If the strategy of adding new error codes for these circumstances is taken for all backends, I would expect many error codes that would only be relevant to very specific circumstances and backends.

I think that this can be made thread-safe with mutex locks. However, I do think there may be circumstances where the wrong message is read for a given error.

Comment (smaslov-intel) [reply sycl/plugins/cuda/pi_cuda.cpp:4760]: I support for adding this new PI API as it makes error-reporting easier () and such will make errors more clear to the end-user in the long run. A somewhat related example is that we are asked to to distinguish between plugin's internal errors vs. underlying native runtime errors. This infrastructure will save us from adding many new PI error codes.

For thread-safety we could block setting a new error until previous message is read (and the last error is cleared and mutex unlocked). This requires that PI_PLUGIN_SPECIFIC_ERROR would have to always be followed by a piPluginGetLastError to avoid dead-locks, but it is easy to document and implement with a wrapper function.

File: sycl/include/CL/sycl/detail/pi.h
Comment (smaslov-intel) [inline sycl/include/CL/sycl/detail/pi.h:1781]: Maybe we can not add "is_warning" but return PI_SUCCESS to indicate that upper RT should treat it as a warning , but return other PI error codes for something that is an error that SYCL RT should try to recover from or throw an exception back to the user.

File: sycl/plugins/cuda/pi_cuda.cpp
Comment (romanovvlad) [reply sycl/plugins/cuda/pi_cuda.cpp:4760]: Another approach to address thread safety issue is to use `thread_local` for messages storage, so each thread can read its own error message after PI call is done.

File: sycl/include/CL/sycl/detail/pi.h
Comment (romanovvlad) [reply sycl/include/CL/sycl/detail/pi.h:1781]: Also we need a way for SYCL RT to know the size of  `message` array. Or we can just return a pointer directly(without the copy) which can be used by the SYCL RT until the next PI call is done(assuming we take `thread_local` approach from another comment).

File: sycl/plugins/cuda/pi_cuda.cpp
Comment (AidanBeltonS) [reply sycl/plugins/cuda/pi_cuda.cpp:4755]: I agree `thread_local` would be a tidy way of handling message storage with multiple threads. I have updated the PR to use it.

File: sycl/include/CL/sycl/detail/pi.h
Comment (AidanBeltonS) [reply sycl/include/CL/sycl/detail/pi.h]: I have updated it to not use `is_warning` but return a user specified code, so PI_SUCCESS is treated as a warning while others will trigger an error. 
Additionally it now uses `thread_local` so it returns a ptr to the message as suggested.

File: sycl/source/detail/plugin.hpp
Comment (romanovvlad) [inline sycl/source/detail/plugin.hpp:118]:
```suggestion
      char *message = nullptr;
```

File: sycl/plugins/cuda/pi_cuda.cpp
Comment (romanovvlad) [inline sycl/plugins/cuda/pi_cuda.cpp:62]:
```suggestion
constexpr size_t MaxMessageSize = 256;
```
?

**Review by romanovvlad — CHANGES_REQUESTED** [review]

File: sycl/source/detail/plugin.hpp
Comment (AidanBeltonS) [reply sycl/source/detail/plugin.hpp]: I have made this change

File: sycl/source/detail/plugin.hpp
Comment (AidanBeltonS) [reply sycl/source/detail/plugin.hpp]: This change is no longer needed. I have undone it

File: sycl/plugins/cuda/pi_cuda.cpp
Comment (AidanBeltonS) [reply sycl/plugins/cuda/pi_cuda.cpp]: This change has been applied

File: sycl/plugins/cuda/pi_cuda.cpp
Comment (AidanBeltonS) [reply sycl/plugins/cuda/pi_cuda.cpp]: No it is not needed any more, I have removed it.

File: sycl/include/CL/sycl/detail/pi.h
Comment (AidanBeltonS) [reply sycl/include/CL/sycl/detail/pi.h]: The description has now been updated.

**Review by romanovvlad — APPROVED** [review]
LGTM

**Comment by AidanBeltonS** [comment]
@smaslov-intel would you be able to re-review the warning level changes?

File: sycl/include/CL/sycl/detail/pi.h
Comment (smaslov-intel) [inline sycl/include/CL/sycl/detail/pi.h:1776]: Please clarify in the API who owns the storage of the message and what it's lifetime is.

**Review by smaslov-intel — APPROVED** [review]
LGTM, just a small ask to clarify error message storage ownership.

File: sycl/include/CL/sycl/detail/pi.h
Comment (AidanBeltonS) [reply sycl/include/CL/sycl/detail/pi.h]: I have clarified the storage and lifetime of the message string

**Review by smaslov-intel — APPROVED** [review]
Thanks!

**Comment by AidanBeltonS** [comment]
@pvchupin are there any other reviewers or is this okay to be merged?

**Pull request merged**
